pkg/client/daemon/outbound_linux.go

```diff
@@ -36,30 +36,51 @@ func (o *outbound) dnsServerWorker(c context.Context) error {
 	return err
 }
 
-func (o *outbound) resolveInSearch(c context.Context, qType uint16, query string) []net.IP {
-	ips := o.resolveInCluster(c, qType, query)
-	if len(ips) == 0 && len(o.search) > 0 {
-		// Apply search path
-		q := query[:len(query)-1]
-		q = strings.ToLower(q)
-		q = strings.TrimSuffix(q, dotTel2SubDomain)
-		if strings.HasSuffix(q, ".local") {
-			return nil
+// applySearch returns true if search path should be applied
+func (o *outbound) applySearch(query string) bool {
+	if len(o.search) == 0 {
+		return false
+	}
+
+	// Don't apply search paths to the kubernetes zone
+	if strings.HasSuffix(query, dotKubernetesZone) {
+		return false
+	}
+
+	// Don't apply search paths if one is already there
+	for _, s := range o.search {
+		if strings.HasSuffix(query, s) {
+			return false
+		}
+	}
+
+	// Don't apply search path to namespaces or "svc".
+	query = query[:len(query)-1]
+	if lastDot := strings.LastIndexByte(query, '.'); lastDot >= 0 {
+		tld := query[lastDot+1:]
+		if _, ok := o.namespaces[tld]; ok || tld == "svc" {
+			return false
 		}
+	}
+	return true
+}
+
+func (o *outbound) resolveInSearch(c context.Context, qType uint16, query string) []net.IP {
+	query = strings.ToLower(query)
+	query = strings.TrimSuffix(query, dotTel2SubDomain)
 
-		q += "."
+	if !o.doClusterLookup(query) {
+		return nil
+	}
+
+	if o.applySearch(query) {
 		for _, s := range o.search {
-			if strings.HasSuffix(q, s) {
-				// Don't append search domain if it's already there
-				continue
-			}
-			ips = o.resolveInCluster(c, qType, q+s)
-			if len(ips) > 0 {
-				break
+			if ips := o.resolveInCluster(c, qType, query+s); len(ips) > 0 {
+				return ips
 			}
 		}
 	}
-	return ips
+	return o.resolveInCluster(c, qType, query)
```

Comment:
I don't get why we have to loop over the search path here.  I understood why we _used to_ have to loop over it; in the pre-`resolveInCluster` days.  But shy are we still doing this client-side, when we can just trim off the trailing `.` and let the agent apply its search path when responding to the LookupHost request?

Reply:
This logic is only used by the overriding resolver. And yes, in most cases it would work to just pass things over. The exception is for local-only intercepts. It should be also be noted that there's a coming limitation that will limit the search path length to one, namely the intercepted namespace so there will be no repeated search.

Reply:
I'm working on a PR to impose the limitation to only be able to intercept one namespace at a time. In that PR, the use of  a search path  is replaced by passing the optional local-only intercepted namespace.